### ISSUE TITLE
AttributeError: module 'bitsandbytes' has no attribute 'nn'

### ISSUE THREAD

**Issue by Harsh-raj** [bitsandbytes-foundation/bitsandbytes#1166]
Why am i getting this error when i am trying to load the checkpoint of lora fine tuned phi-1_5 model. Following is the complete error log
```
phi) harsh@harsh:~/phi-1_5$ python3 inference_lora_phi.py
Traceback (most recent call last):
  File "/home/harsh/phi-1_5/inference_lora_phi.py", line 14, in <module>
    model = PeftModel.from_pretrained(model, "./output/phi-1_5_FT_lr1e-5_ep3_batch8_lora1632_denfc12qkv_proj/checkpoint-14555")
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/harsh/phi-1_5/peft/src/peft/peft_model.py", line 387, in from_pretrained
    model = MODEL_TYPE_TO_PEFT_MODEL_MAPPING[config.task_type](model, config, adapter_name)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/harsh/phi-1_5/peft/src/peft/peft_model.py", line 1212, in __init__
    super().__init__(model, peft_config, adapter_name)
  File "/home/harsh/phi-1_5/peft/src/peft/peft_model.py", line 131, in __init__
    self.base_model = cls(model, {adapter_name: peft_config}, adapter_name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/harsh/phi-1_5/peft/src/peft/tuners/lora/model.py", line 136, in __init__
    super().__init__(model, config, adapter_name)
  File "/home/harsh/phi-1_5/peft/src/peft/tuners/tuners_utils.py", line 165, in __init__
    self.inject_adapter(self.model, adapter_name)
  File "/home/harsh/phi-1_5/peft/src/peft/tuners/tuners_utils.py", line 342, in inject_adapter
    self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)
  File "/home/harsh/phi-1_5/peft/src/peft/tuners/lora/model.py", line 220, in _create_and_replace
    new_module = self._create_new_module(lora_config, adapter_name, target, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/harsh/phi-1_5/peft/src/peft/tuners/lora/model.py", line 282, in _create_new_module
    from .bnb import dispatch_bnb_8bit
  File "/home/harsh/phi-1_5/peft/src/peft/tuners/lora/bnb.py", line 272, in <module>
    if is_bnb_4bit_available():
       ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/harsh/phi-1_5/peft/src/peft/import_utils.py", line 33, in is_bnb_4bit_available
    return hasattr(bnb.nn, "Linear4bit")
                   ^^^^^^
```

Following is the inference code i am using to inference the lora finetuned model

```
import gradio as gr 
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers import StoppingCriteria, StoppingCriteriaList, TextIteratorStreamer
from threading import Thread
from peft import LoraConfig, get_peft_model, PeftModelForCausalLM, PeftConfig, PeftModel

# # Loading the tokenizer and model from Hugging Face's model hub.
tokenizer = AutoTokenizer.from_pretrained("microsoft/phi-1_5", torch_dtype="auto", trust_remote_code=True)
# model = AutoModelForCausalLM.from_pretrained("microsoft/phi-1_5",  trust_remote_code=True)

config = PeftConfig.from_pretrained("./output/phi-1_5_FT_lr1e-5_ep3_batch8_lora1632_denfc12qkv_proj/checkpoint-14555")
model = AutoModelForCausalLM.from_pretrained(config.base_model_name_or_path)
model = PeftModel.from_pretrained(model, "./output/phi-1_5_FT_lr1e-5_ep3_batch8_lora1632_denfc12qkv_proj/checkpoint-14555")
# using CUDA for an optimal experience
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = model.to(device)


# Defining a custom stopping criteria class for the model's text generation.
class StopOnTokens(StoppingCriteria):
    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
        stop_ids = [50256]  # IDs of tokens where the generation should stop.
        for stop_id in stop_ids:
            if input_ids[0][-1] == stop_id:  # Checking if the last generated token is a stop token.
                return True
        return False


# Function to generate model predictions.
def predict(message, history):
    history_transformer_format = history + [[message, ""]]
    stop = StopOnTokens()

    # Formatting the input for the model.
    #messages = "<|endoftext|>".join(["<|endoftext|>".join(["\n" + item[0], "\n" + item[1]])
    #                   for item in history_transformer_format])

    model_inputs = tokenizer([message], return_tensors="pt").to(device) #messages
    streamer = TextIteratorStreamer(tokenizer, timeout=10., skip_prompt=True, skip_special_tokens=True)
    generate_kwargs = dict(
        model_inputs,
        streamer=streamer,
        max_new_tokens=512,
        do_sample=True,
        top_p=0.95,
        top_k=50,
        temperature=0.5,
        num_beams=1,
        stopping_criteria=StoppingCriteriaList([stop])
    )
    t = Thread(target=model.generate, kwargs=generate_kwargs)
    t.start()  # Starting the generation in a separate thread.
    partial_message = ""
    for new_token in streamer:
        partial_message += new_token
        if '<|endoftext|>' in partial_message:  # Breaking the loop if the stop token is generated.
            break
        yield partial_message


# Setting up the Gradio chat interface.
gr.ChatInterface(predict,
                 title="Phi1.5").launch(share=True)  # Launching the web interface.
```

**Comment by matthewdouglas** [comment]
Hi @Harsh-raj,
What version of `bitsandbytes` are you using?

**Comment by Harsh-raj** [comment]
I tried to compile bitsandbytes from source but when i ran the command `cmake -DCOMPUTE_BACKEND=cuda -S .` I am getting this error.
-- Configuring bitsandbytes (Backend: cuda)
-- NO_CUBLASLT := OFF
-- CUDA Version: 101 (10.1.243)
-- CUDA Compiler: /usr/bin/nvcc
CMake Error at CMakeLists.txt:88 (message):
  CUDA Version < 11 is not supported


-- Configuring incomplete, errors occurred!
The system that i am using a bit old. Is there a way to go around this.

**Comment by Harsh-raj** [comment]
So I instead installed bitsandbytes-cuda112-0.26.0.post2 and removed the bitsandbytes source folder.
Now when i try to run the LoRA finetuned model, It throws me this.
```
(phi) harsh@harsh:~/phi-1_5$ python3 inference_lora_phi.py
Traceback (most recent call last):
  File "/home/harsh/phi-1_5/inference_lora_phi.py", line 6, in <module>
    from peft import LoraConfig, get_peft_model, PeftModelForCausalLM, PeftConfig, PeftModel
  File "/home/harsh/phi-1_5/peft/src/peft/__init__.py", line 22, in <module>
    from .auto import (
  File "/home/harsh/phi-1_5/peft/src/peft/auto.py", line 31, in <module>
    from .config import PeftConfig
  File "/home/harsh/phi-1_5/peft/src/peft/config.py", line 23, in <module>
    from .utils import CONFIG_NAME, PeftType, TaskType
  File "/home/harsh/phi-1_5/peft/src/peft/utils/__init__.py", line 21, in <module>
    from .loftq_utils import replace_lora_weights_loftq
  File "/home/harsh/phi-1_5/peft/src/peft/utils/loftq_utils.py", line 35, in <module>
    import bitsandbytes as bnb
  File "/home/harsh/miniconda3/envs/phi/lib/python3.12/site-packages/bitsandbytes/__init__.py", line 5, in <module>
    from .optim import adam
  File "/home/harsh/miniconda3/envs/phi/lib/python3.12/site-packages/bitsandbytes/optim/__init__.py", line 5, in <module>
    from .adam import Adam, Adam8bit, Adam32bit
  File "/home/harsh/miniconda3/envs/phi/lib/python3.12/site-packages/bitsandbytes/optim/adam.py", line 11, in <module>
    from bitsandbytes.optim.optimizer import Optimizer2State
  File "/home/harsh/miniconda3/envs/phi/lib/python3.12/site-packages/bitsandbytes/optim/optimizer.py", line 6, in <module>
    import bitsandbytes.functional as F
  File "/home/harsh/miniconda3/envs/phi/lib/python3.12/site-packages/bitsandbytes/functional.py", line 13, in <module>
    lib = ct.cdll.LoadLibrary(os.path.dirname(__file__) + '/libbitsandbytes.so')
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/harsh/miniconda3/envs/phi/lib/python3.12/ctypes/__init__.py", line 460, in LoadLibrary
    return self._dlltype(name)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/harsh/miniconda3/envs/phi/lib/python3.12/ctypes/__init__.py", line 379, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: libcudart.so.11.0: cannot open shared object file: No such file or directory
```

**Comment by Titus-von-Koeller** [comment]
Hey @Harsh-raj,

Thanks for reporting to us.

First of all, in the title you mention `AttributeError: module 'bitsandbytes' has no attribute 'nn'`, but I don't see that in the error logs that you provided. When did you get this error, the context of that would be important.

Based on the title that I just mentioned, to me the most likely cause of an error sounds like you were inside of the `bitsandbytes` clone directory as your current working directory.

This would mean that Python would look there for the `bitsandbytes` package and at the first level of the hierarchy it doesn't find nn, because that's still one level down.

We really have to switch to a `src` directory instead of using `bitsandbytes`. This is super confusing and has often lead to issues in the past. I'll make that a priority.

Let me know if this helped and otherwise please provide all the details you can, so we can try to help you more.

**Comment by XiuyeYuan** [comment]
Meet the same error with "AttributeError: module 'bitsandbytes' has no attribute 'nn'
##system info:
```
(rocm) root@nv48-rocm:/mnt/myrocm/llama_model# pip show bitsandbytes
Name: bitsandbytes
Version: 0.44.1.dev0+cd3cb68
Summary: k-bit optimizers and matrix multiplication routines.
Home-page: https://github.com/TimDettmers/bitsandbytes
Author: Tim Dettmers
Author-email: dettmers@cs.washington.edu
License: MIT
Location: /root/miniconda3/envs/rocm/lib/python3.10/site-packages
Editable project location: /mnt/myrocm/llama_model/bitsandbytes
Requires: numpy, torch
```
##Error message:

```
Traceback (most recent call last):
  File "/mnt/myrocm/llama_model/fine_tune.py", line 70, in <module>
    model = get_peft_model(base_model, peft_parameters)
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/mapping.py", line 136, in get_peft_model
    return MODEL_TYPE_TO_PEFT_MODEL_MAPPING[peft_config.task_type](model, peft_config, adapter_name=adapter_name)
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/peft_model.py", line 1059, in __init__
    super().__init__(model, peft_config, adapter_name)
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/peft_model.py", line 126, in __init__
    self.base_model = cls(model, {adapter_name: peft_config}, adapter_name)
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/tuners/lora/model.py", line 111, in __init__
    super().__init__(model, config, adapter_name)
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/tuners/tuners_utils.py", line 147, in __init__
    self.inject_adapter(self.model, adapter_name)
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/tuners/tuners_utils.py", line 302, in inject_adapter
    self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/tuners/lora/model.py", line 182, in _create_and_replace
    new_module = self._create_new_module(lora_config, adapter_name, target, **kwargs)
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/tuners/lora/model.py", line 244, in _create_new_module
    from .bnb import dispatch_bnb_8bit
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/tuners/lora/bnb.py", line 218, in <module>
    if is_bnb_4bit_available():
  File "/root/miniconda3/envs/rocm/lib/python3.10/site-packages/peft/import_utils.py", line 31, in is_bnb_4bit_available
    return hasattr(bnb.nn, "Linear4bit")
AttributeError: module 'bitsandbytes' has no attribute 'nn'

```

**Comment by s-zk** [comment]
Install bitsandbytes from source follow https://github.com/bitsandbytes-foundation/bitsandbytes/blob/main/docs/source/installation.mdx#cuda-compile

after `make` use `pip install .` instead of `pip install -e .`

works for me.

**Comment by kawakuboJin** [comment]
> Install bitsandbytes from source follow https://github.com/bitsandbytes-foundation/bitsandbytes/blob/main/docs/source/installation.mdx#cuda-compile
> 
> after `make` use `pip install .` instead of `pip install -e .`
> 
> works for me.

brilliant! it works

**Comment by mehdiae** [comment]
I am facing the same issue for AMD Rocm. After building and pip installation I faced this issue:
```
File /opt/conda/envs/llms/lib/python3.12/site-packages/transformers/integrations/bitsandbytes.py:199, in 
_replace_with_bnb_linear(model, modules_to_not_convert, current_key_name, quantization_config, has_been_replaced)
    195     pass
    196 else:
    197     extra_kwargs = (
    198         {"quant_storage": quantization_config.bnb_4bit_quant_storage}
--> 199         if "quant_storage" in list(signature(bnb.nn.Linear4bit).parameters)
    200         else {}
    201     )
    202     model._modules[name] = bnb.nn.Linear4bit(
    203         in_features,
    204         out_features,
   (...)
    209         **extra_kwargs,
    210     )
    211     has_been_replaced = True

AttributeError: module 'bitsandbytes' has no attribute 'nn'
```

**Comment by TimDettmers** [comment]
Closing this issue. The `AttributeError: module 'bitsandbytes' has no attribute 'nn'` error was a symptom of the C/CUDA library failing to load on older bitsandbytes versions — when the library doesn't load, the module initialization is incomplete. The various commenters here report different root causes (wrong version, ROCm build issues, etc.) but the common thread is old bnb versions with poor error reporting. Library loading and error messaging were overhauled in v0.43.0+ and further improved by PR #1615.

If you're still hitting this on the **latest** bitsandbytes (v0.45+), please open a new issue with the output of `python -m bitsandbytes` and your environment details.
